### PR TITLE
3 header

### LINKED ISSUE #3: Header
Implement header based on new designs for all screen sizes. Search functionality is out of scope for this issue (search icon points to the search page).

https://www.figma.com/file/IcqreRKnNYyDVMc28yoOeh/Homepage?type=design&node-id=4271-21750&mode=design&t=7R9SLcKCinr5XmCO-4

- The links will follow the [default/selected/hover state from Volvo Core Design Kit](https://www.figma.com/file/JglsLWtLIgdFru0p8Va2vz/BETA---Volvo-Core-Design-Kit?type=design&node-id=700-29717&mode=design&t=qiy9Hsxnz6BB3jKC-4). The focus state will be the same as the hover state.

### Requirements
- On desktop, an extra link is shown ("Find a dealer" with icon)
- On mobile and tablet, the links are placed in a modal side menu when the user taps the hamburger menu icon

#### Behaviour
- When the header is out of view, and the user scrolls up, the header will appear (sticky). It will be hidden [once the user scrolls down again](https://bootstrap-menu.com/demos/autohide.html).
- 

#### Styling
- The links will follow the [default/selected/hover state from Volvo Core Design Kit](https://www.figma.com/file/JglsLWtLIgdFru0p8Va2vz/BETA---Volvo-Core-Design-Kit?type=design&node-id=700-29717&mode=design&t=qiy9Hsxnz6BB3jKC-4). The focus state will be the same as the hover state.
- Menu items in main navigation are centre-aligned
- On mobile and tablet, the following motion will happen:
  - Appearing:
    - Move sideways into the screen
    - Duration: 450ms
    - Easing: easing-entrance
  - Disappearing:
    - Move sideways out of the screen
    - Duration: 450ms
    - Easing: easing-exit
- When the user has set in their OS they prefer reduced motion, the animations will be disabled

image

#### Icons
The following [icons](https://www.volvogroup.com/experiencesystem/en/#/brand/615bfcc428c818001105b4e3/61a0de32a7047974164f01dd/622b4b1737a675408d76d87b) should be used:

- Control/Close
- Control/Search
- Navigation/Hamburger Menu
- Navigation/Chevron Down
- Navigation/Chevron Up
- People/User
- Vehicles/Truck NA

#### Open points (relate to a follow up ticket for navigation):
- Burger menu content
- Fold out menu for each navigation item to be defined
- Services and about volvo are not direct links, but trucks and news might be direct links

### PR DESCRIPTION
**Pull request by TomaszDziezykNetcentric**
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #3 

Test URLs:
- Before: https://main--vg-volvotrucks-us-rd--netcentric.hlx.page/
- After: https://3-header--vg-volvotrucks-us-rd--netcentric.hlx.page/

### CODE REVIEW

**Review by cogniSyb — CHANGES_REQUESTED** [review]

**Review by cogniSyb — CHANGES_REQUESTED** [review]

File: blocks/header/header.css [hunk 1]
```diff
@@ -1,809 +1,265 @@
...
+  gap: 24px;
 }
 
-header nav .hamburger-toggle[aria-expanded='true'] ~ .sections {
-  right: 0;
-}
-
-/* section entries */
-header nav .sections ul {
+.header__actions-list {
```
Comment (kesiah) [inline blocks/header/header.css:121]: Lots of styles are shared btw `.header__actions-list` and `.header__main-links`, can we put them once?

File: blocks/header/header.js [hunk 2]
```diff
@@ -1,306 +1,198 @@
...
+      href="#"
+      aria-label="Search"
+      class="${blockClass}__search-button ${blockClass}__action-link ${blockClass}__link"
+    >
+      <span class="icon icon-search-icon"></span>
+      <span class="${blockClass}__search-label">Search</span>
+    </a>
+    <a
+      href="#"
+      aria-label="Open menu"
```
Comment (kesiah) [inline blocks/header/header.js:94]: Placeholders?

**Review by kesiah — CHANGES_REQUESTED** [review]
If the text in the menu is a bit larger, everything goes in 2 lines. Should we allow the menu to grow? There is a lot of empty space that can be used.

image

File: icons/hamburger-icon.svg [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="56" height="57" viewBox="0 0 56 57" fill="none">
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply icons/hamburger-icon.svg]: fixed

File: blocks/header/header.css [hunk 4]
```diff
@@ -1,809 +1,265 @@
...
-
-header nav .tools ul li a {
-  color: var(--c-grey-4);
-  text-decoration: none;
-  font-size: 8pt;
+  gap: 4px;
+  color: var(--c-grey-700);
+  font-feature-settings: 'clig' off, 'liga' off;
+  font-family: var(--ff-volvo-novum);
+  font-size: 12px;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css]: fixed

File: blocks/header/header.css [hunk 5]
```diff
@@ -1,809 +1,265 @@
...
-  header nav .sections .section.expand > ul > li > p.title {
-    font-size: 24px;
-    line-height: 1.333;
-    border-bottom: 1px solid var(--background-light-gray);
-    padding-bottom: 10px;
-    margin-bottom: 8px;
+  .header__main-links {
+    height: auto;
+    position: static;
+    padding: 0 16px;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css:224]: fixed

File: blocks/header/header.css [hunk 6]
```diff
@@ -1,809 +1,265 @@
...
-  flex-flow: column;
-}
+  justify-content: center;
 
-/* volvo group section */
-header nav .brand .vgsection-location .vgsection {
-  font-family: var(--font-family-body);
-  color: var(--text-color);
-  font-weight: 700;
+  --menu-animation: right 450ms ease-in-out;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css:5]: Don't know I implement it as it is on the design. If we want to change it we need to ask the agency, right? @cogniSyb ?

File: blocks/header/header.css (hunk 1 above)
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css]: Yup right, fixed

File: blocks/header/header.js [hunk 7]
```diff
@@ -1,306 +1,198 @@
...
+
+        textNode.replaceWith(spanWrapper);
+        spanWrapper.append(textNode);
+        link.setAttribute('aria-label', textNode.textContent);
+      });
+  });
+
+  const closeIcon = document.createRange().createContextualFragment(`
+    <button
+      aria-label="Close menu"
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js]: Yes it should, and now it is there ;)

File: blocks/header/header.js (hunk 2 above)
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js]: Yup, fixed

File: blocks/header/header.js [hunk 8]
```diff
@@ -1,306 +1,198 @@
...
+      <span class="icon icon-hamburger-icon"></span>
+    </a>
+  `);
+
+  mobileActionsEl.append(...actions.childNodes);
+
+  return mobileActionsEl;
+};
+
+const addHeaderScrollBehavtios = (header) => {
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js]: Nice catch, fixed

File: styles/styles.css [hunk 9]
```diff
@@ -254,7 +254,7 @@
   --body-regular-letter-spacing: 0.5px;
 
   /* nav height */
-  --nav-height: 95px;
+  --nav-height: 52px;
   --sub-nav-height: 44px;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply styles/styles.css:258]: This is the topic for discussion @cogniSyb  @Lakshmishri.
Before we had sub-nav, but I don't know if the sub-nav will be on the other pages - that are out of scope for the redesign for now. So, we should have an issue with removing the sub nav block (and docs) or we should check how it is displayed with the new header (and do alignment if needed).

File: styles/styles.css [hunk 10]
```diff
@@ -854,13 +853,6 @@ div.block div.pager .pagination .scroll li a {
...
 
 @media (max-width: 991px) {
-  header .header.block {
-    position: fixed;
-    top: 0;
-    left: 0;
-    right: 0;
-  }
-
   header .sub-nav.block {
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply styles/styles.css:856]: As above - topic to discuss

File: blocks/header/header.js [hunk 11]
```diff
@@ -1,306 +1,198 @@
...
+  const mobileActionsEl = createElement('div', { classes: [`${blockClass}__mobile-actions`] });
+
+  const actions = document.createRange().createContextualFragment(`
+    <a
+      href="#"
+      aria-label="Search"
+      class="${blockClass}__search-button ${blockClass}__action-link ${blockClass}__link"
+    >
+      <span class="icon icon-search-icon"></span>
+      <span class="${blockClass}__search-label">Search</span>
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js]: fixed

File: styles/styles.css [hunk 12]
```diff
@@ -295,7 +295,6 @@ header {
 header .header.block {
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply styles/styles.css]: Updated

File: icons/search-icon.svg
Comment (Lakshmishri) [inline icons/search-icon.svg:1]: Do we need the id for `<g>` ?

File: blocks/header/header.css [hunk 13]
```diff
@@ -1,809 +1,268 @@
...
+  width: 100vw;
+  height: 100vh;
+  display: block;
 }
 
-header nav .sections .section > ul > li.expand > ul li a::after {
-  content: unset;
+.header--hamburger-open .header__actions-list {
+  right: 0;
+  top: 0;
```
Comment (Lakshmishri) [inline blocks/header/header.css:177]: You can remove `top:0 ` have same css rule at 132

File: blocks/header/header.css [hunk 14]
```diff
@@ -1,809 +1,268 @@
...
-    width: 100%;
-    z-index: 1010;
   }
 
-  header nav .sections .section.expand .navigation-cta {
-    margin: 0 0 0 -92px !important;
-    width: calc(100% + 92px) !important;
-    padding: 16px 0;
+  .header__actions-list,
+  .header__main-links {
```
Comment (Lakshmishri) [inline blocks/header/header.css:218]: Do we need this ? I see that `display:flex` is set at 122

File: blocks/header/header.js [hunk 15]
```diff
@@ -1,306 +1,200 @@
...
+  const actions = document.createRange().createContextualFragment(`
+    <a
+      href="#"
+      aria-label="${searchLable}"
+      class="${blockClass}__search-button ${blockClass}__action-link ${blockClass}__link"
+    >
+      <span class="icon icon-search-icon"></span>
+    </a>
+    <button
+      href="#"
```
Comment (kesiah) [inline blocks/header/header.js:95]: Please remove href

**Review by Lakshmishri — DISMISSED** [review]
Screenshot 2023-09-05 at 09 37 08

- Can you align the border bottom with rest of them (above attached screen shot)
- Currently the dealer locator is pointing to search page can you change it to dealer locator page?
- Can you check scroll behaviour in Safari , it's jumpy and doesn't show header when you scroll back to top of the page

**Review by kesiah — CHANGES_REQUESTED** [review]
Check the fonts, because I see some inconsistencies while comparing with the designs.
Screenshot 2023-09-05 at 09 41 28

In desktop:
- Menu items should use: --f-button-font-size
- Action items should use: --f-button-font-size

In mobile:
- Menu items should use: --f-body-font-size
- Action items should use: --f-caption-font-size

Make sure to use the font size variable, but also line-height and letter-spacing ;)

Apart from this, the border is not following the design, compare the screenshot (marked in yellow).

**Review by cogniSyb — CHANGES_REQUESTED** [review]

**Comment by kesiah** [comment]
@cogniSyb  `--nav-height` can be necessary for the inpage navigation, we need to provide a scroll-margin to the content so it's scrolled to a good position.

File: blocks/header/header.js [hunk 16]
```diff
@@ -1,306 +1,200 @@
...
-    // write the section menu
-    if (sectionMenus.length) {
-      sectionList.append(...sectionMenus);
-    }
+  // get the navigation text, turn it into html elements
+  const content = document.createRange().createContextualFragment(await resp.text());
+  const [logoContainer, navigationContainer, actionsContainer] = [...content.querySelectorAll('div')];
+  const nav = createElement('nav', { classes: [`${blockClass}__nav`] });
+  const navContent = document.createRange().createContextualFragment(`
+    <div class="${blockClass}__menu-background"></div>
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js:148]: This is the grey background below the menu. When the navigation is on the right the rest of the page has grey background layer. 
obraz

File: icons/search-icon.svg
Comment (TomaszDziezykNetcentric) [reply icons/search-icon.svg:1]: Removed

File: blocks/header/header.css (hunk 13 above)
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css]: Removed

File: blocks/header/header.css (hunk 14 above)
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css]: Yup, it is not needed anymore - removed

File: blocks/header/header.js (hunk 15 above)
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js]: done

File: blocks/header/header.js [hunk 17]
```diff
@@ -1,306 +1,200 @@
...
+      aria-label="${searchLable}"
+      class="${blockClass}__search-button ${blockClass}__action-link ${blockClass}__link"
+    >
+      <span class="icon icon-search-icon"></span>
+    </a>
+    <button
+      href="#"
+      aria-label="${openMenuLable}"
+      class="${blockClass}__hamburger-menu ${blockClass}__action-link ${blockClass}__link"
+      aria-expanded="false"
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js:97]: I updated the code. I used the `aria-hidden`. When the nav is hidden the items should be ignored.
obraz

File: blocks/header/header.css [hunk 18]
```diff
@@ -1,809 +1,268 @@
...
-    border: 0;
-    padding: 15px 24px;
-    height: 48px;
-    cursor: pointer;
+    justify-content: center;
   }
 
-  header nav .sections .section.expand {
+  .header__nav {
     padding: 0;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css]: done

File: blocks/header/header.css [hunk 19]
```diff
@@ -1,809 +1,268 @@
...
-    width: 50vw;
-    right: -50vw;
-  }
-
-  header nav .hamburger-toggle[aria-expanded='true'] ~ .sections {
-    width: 50vw;
-    right: 0;
-  }
+.header--hidden {
+  top: calc(-1 * var(--nav-height));
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css]: done

**Comment by TomaszDziezykNetcentric** [comment]
> Check the fonts, because I see some inconsistencies while comparing with the designs. Screenshot 2023-09-05 at 09 41 28
> 
> In desktop:
> 
>     * Menu items should use: --f-button-font-size
> 
>     * Action items should use: --f-button-font-size
> 
> 
> In mobile:
> 
>     * Menu items should use: --f-body-font-size
> 
>     * Action items should use: --f-caption-font-size
> 
> 
> Make sure to use the font size variable, but also line-height and letter-spacing ;)
> 
> Apart from this, the border is not following the design, compare the screenshot (marked in yellow).

I used the line height and letter-spacing, but I can use the font-size value because they are calculated based on the doc font size which is different between the redesign and other pages. If we use the font-size variable, the pages that are out of the redesign scrope will have smaller font sizes because of the CSS rule: 
obraz

**Review by cogniSyb — CHANGES_REQUESTED** [review]

File: blocks/header/header.js [hunk 20]
```diff
@@ -1,306 +1,205 @@
...
+    link.classList.add(`${blockClass}__main-nav-link`, `${blockClass}__main-nav-link--collapsed`, `${blockClass}__link`);
 
-  if (navCta) {
-    const li = document.createElement('li');
-    li.className = 'navigation-cta';
-    li.innerHTML = navCta.innerHTML;
-    subSectionMenus.push(li);
-  }
+    const chevronIcon = createElement('span', { classes: ['icon', 'icon-chevron-down'] });
+    link.append(chevronIcon);
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js]: done

File: blocks/header/header.js [hunk 21]
```diff
@@ -1,306 +1,205 @@
...
-      a.addEventListener('click', toggleSectionMenu.bind(a, sectionMenu, navCta, menuBlock));
-      sectionMenu.appendChild(a);
-      return sectionMenu;
+  // get the navigation text, turn it into html elements
+  const content = document.createRange().createContextualFragment(await resp.text());
+  const [logoContainer, navigationContainer, actionsContainer] = [...content.querySelectorAll('div')];
+  const nav = createElement('nav', { classes: [`${blockClass}__nav`] });
+  const navContent = document.createRange().createContextualFragment(`
+    <div class="${blockClass}__menu-background"></div>
+    <div class="${blockClass}__logo-wrapper">
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js]: `<div>` removed

File: blocks/header/header.js [hunk 22]
```diff
@@ -1,306 +1,205 @@
...
-    decorateIcons(nav);
-    /* append result */
-    block.append(nav);
-    loadSearchWidget();
-  }
+    setAriaForMenu(true);
+  });
+
+  navContent.querySelectorAll(`.${blockClass}__menu-background, .${blockClass}__close-menu`).forEach((el) => {
+    el.addEventListener('click', closeHamburderMenu);
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js]: Typo fixed

File: blocks/header/header.css [hunk 23]
```diff
@@ -1,809 +1,282 @@
...
+  left: 0;
+  background: rgb(20 20 20 / 40%);
+  width: 100vw;
+  height: 100vh;
+  display: block;
 }
 
-header nav .sections .section > ul > li.expand a::after {
-  content: '\f068';
+.header--hamburger-open .header__actions-list {
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css:199]: What do you mean by misaligned? I moved the hamburger menu so all of the icons are now in the same line.

File: styles/styles.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply styles/styles.css:260]: This is the preview of the page with subnav: https://3-header--vg-volvotrucks-us-rd--netcentric.hlx.page/innovation/supertruck/ (On mobile the page content is broken but the headed and sub nav work as should)

**Review by cogniSyb — CHANGES_REQUESTED** [review]

File: blocks/header/header.js [hunk 24]
```diff
@@ -1,306 +1,200 @@
...
+      aria-label="${closeMenuLabel}"
+      class="${blockClass}__close-menu"
+      aria-expanded="false"
+      aria-controls="header-main-nav, header-actions-list"
+    >
+      <span class="icon icon-close" />
+    </button>
+  `);
+
+  list.append(closeIcon);
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.js:75]: done

**Review by cogniSyb — APPROVED** [review]

**Review by kesiah — CHANGES_REQUESTED** [review]
I cannot still see the updates on accessibility, for example when the menu in mobile is expanded, the hamburger button should update the property aria-expanded.

File: blocks/header/header.css [hunk 25]
```diff
@@ -1,809 +1,290 @@
...
-    margin: 25px 0 0;
-    padding: 0;
+    justify-content: center;
   }
 
-  /* hide mobile search toggle */
-  header nav .search-toggle {
-    display: none;
+  .header__nav {
+    max-width: 1232px;
```
Comment (kesiah) [inline blocks/header/header.css:223]: Please use `var(--wrapper-width)` so all the items in the website have the same width

**Review by kesiah — CHANGES_REQUESTED** [review]

File: blocks/header/header.css (hunk 25 above)
Comment (TomaszDziezykNetcentric) [reply blocks/header/header.css]: Updated to `max-width: calc(var(--wrapper-width) + 32px);`

**Comment by TomaszDziezykNetcentric** [comment]
> I cannot still see the updates on accessibility, for example when the menu in mobile is expanded, the hamburger button should update the property aria-expanded.

I had a typo in selectors :( 
Fixed

File: blocks/header/header.css (hunk 25 above)
Comment (kesiah) [reply blocks/header/header.css]: Problem is that then the in page navigation needs to behave the same :(
If not, they look weird.

**Review by kesiah — APPROVED** [review]

**Pull request merged**